### ISSUE TITLE
Implicit json casting can cause confusion

### ISSUE THREAD

**Issue by eladcon** [winglang/wing#4771]
### Feature Spec

The compiler now enforces a more strict type checking when working with Json objects. 

For example, the Json representation of a wing struct must be an object Json and so the method `fromJson` on structs accept only those kind of types. So the following are now compiler errors:
```js
MyStruct.fromJson(false);
MyStruct.fromJson(Json false);
MyStruct.fromJson("a string");
MyStruct.fromJson(Json "a string");
MyStruct.fromJson([1,2,3]);
MyStruct.fromJson(Json [1,2,3]);
```

### Use Cases

see above

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

**Comment by Chriscbr** [comment]
To support this kind of feature I think we would need to add more complexity to how we model JSON in Winglang's type system. Today, the signature of `MyStruct.fromJson` accepts a `Json` type, but we don't have a way in Wing code to express a function only expects certain variants of JSON.

If we modeled `Json` as something like this (pseudocode):

```js
type Json = JsonBool | JsonNum | JsonStr | JsonArray | JsonObj | JsonNull;
// each variant is a subtype of Json
```

Then we could update the signature of "fromJson" to look like this:
```js
static fromJson(obj: JsonObj): MyStruct;
```

But I'm not sure if this might add confusion in other ways (e.g. what's the difference between `JsonStr` and `str`)?

**Comment by hasanaburayyan** [comment]
Hey @eladcon Im curious, I understand why this fails and what the issue is. But Im curious about the use case here, what does parsing `MyStruct.fromJson(true)` enable? 

I.E. What does `MyStruct` look like. Technically we could get past the compilation errors using `unsafe` mode like:
```ts
MyStruct.fromJson(true, { unsafe: true })` 
```

but this would wind up assigning the value `true` to the variable and in essence no longer be a struct.

**Comment by eladb** [comment]
@eladcon the use case you ran into based on our conversation was parsing JSON strings into structs.

Let's focus on that use case because I don't think adding more fine grained json typing is something we will be able to address right now.

I suggest that for now, we will just add a bunch of static methods `parseJson(str?)` (and `tryParseJson(str?)`) to each struct.

We want to optimize our API ergonomics for the happy path of receiving an API response with an expected structure and using it in my code:

```js
let response = http.get(...);
let foo = Foo.parseJson(response.body);
log(foo.bang);
```

Let's get all of our `.parse()` methods (I think we have `Json.parse`) to also accept optional strings (and fail if they are `nil`).

@staycoolcall911 we can repurpose this issue to address this use case or open another issue.

**Comment by Chriscbr** [comment]
@eladb respectfully I'd like to ask whether we can reconsider changing all parse methods to accept optional strings. In my opinion it feels like poor API design to widen the allowed types of a function to allow more invalid inputs to be given.

Regarding `Json.parse`, by allowing `nil` as an argument, we're introducing ambiguity into parsing that can mask errors stemming from upstream data handling. For instance, suppose an unexpected `nil` is passed due to a bug in an earlier function:

```js
let complexResult = complexFunction(); // expected to return str, erroneously returns str?
let jsonData = Json.parse(complexResult); // throws, diverting attention from the real issue
```

Before the API change, this would be a compile time error -- so if `complexFunction` returned `nil`, the developer would have to explicitly handle this case, possibly by throwing an error, logging a warning, or retrying the operation. Alternatively, the developer might discover that `complexFunction` shouldn't have been returning nil, and could fix the code in the first line.

After the API change, this would not be a compiler time error -- so if `complexFunction` returned `nil`, the code would error, but the stack trace would erroneously suggest that parsing was the source of the problem. Likewise, if the above example used `Json.tryParse` instead, then the `nil` result that would follow would likely be interpreted as "the data is not JSON" and not "there was no data to parse", since the error is silently discarded.

More broadly, I think altering `Json.parse` to throw a runtime exception on a nil argument also increases the cognitive load on the developer to remember to handle nil checks preemptively. Indeed, even TypeScript treats passing `undefined` and `null` to `JSON.parse()` as a comple-time error ([playground](https://www.typescriptlang.org/play?#code/FIZQ8gcgdADghgJwM4FMAUBXAdgExQMwEssUcBKAbgFgAoUSWRVNLDAGzctqA)).

----

> We want to optimize our API ergonomics for the happy path of receiving an API response with an expected structure and using it in my code.

As an alternative for improving the ergonomics here, how about we change `body` returned by `util.fetch()` to return a non-optional `str`? From what I could find, many languages today like TypeScript, Kotlin, and Golang model HTTP response bodies in this way. If a developer wants to be more precise, they can check whether the body is empty or if there's no body by checking if the response `status` code is 204 or 200 explicitly.

**Comment by eladb** [comment]
> @eladb respectfully I'd like to ask whether we can reconsider changing all parse methods to accept optional strings. In my opinion it feels like poor API design to widen the allowed types of a function to allow more invalid inputs to be given.

No problem, let's start with non-optional for `parse(str)` and optional for `tryParse(str?)` and see how it works.

**Comment by eladb** [comment]
>  i still think some of the confusion comes from the implicit conversions around Json.

@eladcon fair enough, I am retitling this issue to represent the problem instead of the solution and we can keep it open and discuss further.

@hasanaburayyan can you open an issue to add the parse methods to structs?

**Comment by hasanaburayyan** [comment]
> @hasanaburayyan can you open an issue to add the parse methods to structs?

@eladb done here it is: https://github.com/winglang/wing/issues/4811

**Comment by Chriscbr** [comment]
One proposed idea - limit implicit JSON casting to within `Json` object / array literals. For example, the following code would still be ok (same as the behavior we have today):

```js
let b: bool = true;
let n: num = 42;

// the expressions `b` and `n` are inside an Json object/array literals,
// so they are converted to `Json` implicitly
let obj1 = { a: b, m: n };
let obj2 = Json ["cool", b, n];
```

But outside of Json literals, the casting would be disallowed (which only means you need to put a `Json` in front to convert it):

```js
let b: bool = true;
let n: num = 42;

let f = printJson(x: Json) => {
  log(Json.stringify(x));
};
f(b); // error: expected type "Json" but received "bool". hint: cast a bool as Json using `Json b`
f(n); // error: expected type "Json" but received "num". hint: cast a num as Json using `Json n`
```
